### ISSUE TITLE
Parallelism level in `ZStream#mapZIOPar` is bounded by the buffer size

### ISSUE THREAD

**Issue by kyri-petrou** [zio/zio#9339]
Currently, the parallelism level in `ZStream#mapZIOPar` is effectively bounded by the optional `bufferSize` argument (defaults to 16). This is implementation detail that comes from how the internal buffer is used to guarantee ordering of the items processed in parallel.

This is not clear to users at all, as they expect the provided parallelism level to be honoured (as reported by #9308). We need to either reimplement `ZStream#mapZIOPar` so that the parallelism value provided is no longer bounded by the `bufferSize` value, or that we provide a more sensible default when the `bufferSize` is not explicitly provided. Note that we cannot simply use `bufferSize = parallelism` because in the past we had reports of this leading to OOM errors for users that used `Int.MaxValue` as the parallelism level.

**Comment by asr2003** [comment]
/attempt #9339

<div id="algora-attempt" />

| [Algora profile](https://console.algora.io/@/asr2003) | Completed bounties | Tech | Active attempts | Options |
| --- | --- | --- | --- | --- |
| @asr2003 | &nbsp;&nbsp;&nbsp;4 ZIO bounties <br/> + 5 bounties from 3 projects | <div align="center">Rust, Scala, <br />Go & more</div> | <div align="center"></div> | [Cancel attempt](https://console.algora.io/api/bounties/cm424yqz70006jw035zf6qbpu/cancel-attempt) |

**Comment by joroKr21** [comment]
You are anyway bounded by the minumum of both parameters if you wait for the `queue.offer`:
```scala
            for {
              f <- permits
                     .withPermit(
                       latch.succeed(()) *> f(outElem)
                         .catchAllCause(cause =>
                           failure.update(_ && cause).unless(cause.isInterruptedOnly) *>
                             errorSignal.succeed(()) *>
                             ZChannel.failLeftUnit
                         )
                     )
                     .interruptible
                     .forkIn(childScope)
              _ <- latch.await
              _ <- outgoing.offer(f)
            } yield ()
```

Why do you even need a `bufferSize` parameter if you use a semaphore? Just use an unbounded queue.

**Comment by regiskuckaertz** [comment]
> Why do you even need a bufferSize parameter if you use a semaphore? Just use an unbounded queue.

💯 these two things are orthogonal but currently it is impossible to opt-out of buffering. Buffering should be opt-in:
```
stream.mapZIOPar(nworkers)(f).buffer(bsize)
```
I don't know if an unbounded queue is the best choice as we would lose the back-pressure semantics of `ZStream`. A variant of `MVar` maybe? Many stream operators transform the upstream into a "pull" IO anyway.

**Comment by joroKr21** [comment]
> I don't know if an unbounded queue is the best choice as we would lose the back-pressure semantics of ZStream. A variant of MVar maybe? Many stream operators transform the upstream into a "pull" IO anyway.

 - Doesn't the semaphore provide back-pressure?
 - If you want unlimited parallelism, that's incompatible with back-pressure.

Alternatively you could use a resizable bounded queue.
The problem with the bounded queue in ZIO is that it allocates the entire space eagerly.

**Comment by regiskuckaertz** [comment]
the semaphore in my mind only controls access to resources used by `f`, but in the face of a slow consumer, `nworkers` workers would keep pulling elements and add them to the unbounded queue. did I get this right?

**Comment by joroKr21** [comment]
Ah yeah, you're right - the semaphore protects the parallelism, once a fiber is completed, the permit will be released. And it wouldn't be put into the queue before it started. But anyway, there's no way to have back-pressure if you want to run all items in the stream in parallel:
>  Note that we cannot simply use bufferSize = parallelism because in the past we had reports of this leading to OOM errors for users that used Int.MaxValue as the parallelism level.

So that means that the queue should be both bounded and resizable.

**Comment by joroKr21** [comment]
> And it wouldn't be put into the queue before it started.

I meant to say that there's no point to set parallelism > buffer size.
If you can't start the next task because the previous one is waiting for a slot in the queue.
Now, there might be a reason to set buffer size > parallelism to accommodate slow consumers.
But, as you mentioned you could just use `buffer(bsize)` if you wanted.
It's a minor optimization to use only one queue instead of two.
But the default should really be parallelism = buffer size.

**Comment by krapcys1-maker** [comment]
Hi! Before attempting a solution, I’d like to clarify expectations.

From the discussion it sounds like:
- the current behavior (parallelism bounded by bufferSize) is considered incorrect,
- but simple fixes (e.g. dynamic buffer sizing) were rejected as band-aids.

Questions:
1) Is the intended direction to *decouple buffering from parallelism* entirely
   (i.e. mapZIOPar controls only concurrency, buffering becomes explicit via `.buffer`)?
2) Are breaking changes to mapZIOPar’s signature acceptable if properly documented?
3) Would an initial PR that focuses on correctness + tests (without final API decision)
   be acceptable for review?

Thanks!

**Comment by BISHT-CX** [comment]
/attempt #9339

### Implementation Plan

The root cause appears to be that `mapZIOPar` uses a bounded queue (`bufferSize`) to maintain ordering, which unintentionally limits effective parallelism to `min(n, bufferSize)` because `queue.offer` blocks the producer loop.

#### Proposed Approach

1. **Decouple parallelism from buffering**

   * Use a `Semaphore(n)` as the single source of truth for concurrency control
   * Ensure that `n` permits strictly define the maximum number of concurrent fibers

2. **Remove artificial parallelism cap**

   * Replace the bounded queue used for in-flight fibers with an unbounded queue
   * This avoids blocking the producer loop before all `n` fibers can be started

3. **Reorder operations to avoid premature backpressure**

   * Change from:

     ```
     latch.await *> outgoing.offer(fiber)
     ```

     to:

     ```
     outgoing.offer(fiber) *> latch.await
     ```
   * This ensures ordering is preserved while preventing the queue from limiting concurrency

4. **Preserve backpressure semantics**

   * Backpressure will be enforced via permit acquisition (`latch.await`)
   * Upstream pulling will be gated by available permits, preventing unbounded accumulation

5. **Maintain ordering guarantees**

   * Fibers will still be enqueued in the same order as upstream elements

6. **Handle `bufferSize`**

   * Retain parameter for binary compatibility
   * Deprecate it and document that buffering should be handled explicitly via `.buffer`

#### Tests

* Verify that `mapZIOPar(n, bufferSize < n)` achieves full parallelism
* Add regression test for the original bug
* Add tests for:

  * `mapZIOParUnordered`
  * partial consumption (`take(1)`)
  * ordering guarantees
  * large parallelism values (safety)

#### Safety Considerations

* Although the queue is unbounded, its size is effectively bounded by `n` due to permit gating
* No eager allocation is introduced, avoiding OOM risks seen in prior approaches

Happy to adjust the approach based on feedback before implementation.

**Comment by beatsandcodes** [comment]
I’d like to work on this issue.

/attempt #9339

From my understanding, the current behavior of `ZStream#mapZIOPar` limits effective parallelism to `min(parallelism, bufferSize)` due to internal buffering required to preserve ordering. This is not obvious to users and can lead to confusion when higher parallelism values are not reflected in execution.

### Proposed approach

I propose to keep the existing ordering guarantees intact, while improving how the default `bufferSize` is determined when it is not explicitly provided:

* Replace the fixed default (`16`) with a value derived from `parallelism`
* Use a bounded strategy to avoid excessive memory usage:

  ```
  effectiveBufferSize = min(parallelism, DEFAULT_MAX_BUFFER)
  ```
* Only apply this logic when `bufferSize` is not explicitly specified by the user (existing explicit behavior remains unchanged)

### Notes on DEFAULT_MAX_BUFFER

* Introduce a reasonable upper bound (e.g. 64 or 128) to prevent OOM risks when very large parallelism values are used (e.g. `Int.MaxValue`)
* The exact value can be aligned with existing conventions in ZIO (if any), or discussed with maintainers

### Why this approach

* Preserves ordering semantics and backward compatibility
* Makes effective parallelism more aligned with user expectations
* Avoids unbounded memory growth
* Keeps the implementation change minimal and predictable

### Optional follow-up (not in this PR)

A potential longer-term improvement could be introducing an unordered variant (e.g. `mapZIOParUnordered`) to fully decouple concurrency from ordering constraints.

**Comment by samjay8** [comment]
/attempt #9339

## Implementation Plan

### Problem Summary

`ZStream#mapZIOPar` currently limits effective parallelism to:

```
min(parallelism, bufferSize)
```

because `queue.offer` blocks when the buffer is full, preventing new fibers from starting.

This violates user expectations that `parallelism` controls concurrency.


## Goal

Decouple:

* **Parallelism (concurrency control)**
  from
* **Buffering (ordering + backpressure)**

while preserving:

* Ordering guarantees
* Backpressure semantics
* Memory safety


## Proposed Approach

### 1. Make Semaphore the Source of Truth

* Use `Semaphore(parallelism)` to strictly control concurrency
* Ensure **up to `parallelism` fibers are always allowed to run**


### 2. Prevent Queue from Blocking Fiber Creation

Current issue:

```scala
latch.await *> outgoing.offer(fiber)
```

Fix:

```scala
outgoing.offer(fiber) *> latch.await
```

Effect:

* Fiber is enqueued immediately
* Queue no longer blocks parallelism
* Ordering preserved via `latch`


### 3. Keep Queue Bounded (Do NOT make it unbounded)

Reject:

* Fully unbounded queue → breaks backpressure

Instead:

* Keep bounded queue
* But ensure:

  * It does **not gate fiber start**
  * Only affects downstream consumption


### 4. Decouple Execution from Emission

Split responsibilities:

* **Execution layer**

  * Controlled by semaphore
  * Starts up to `parallelism` fibers

* **Emission layer**

  * Controlled by queue
  * Preserves ordering + backpressure


### 5. Maintain Ordering via Latch / Indexing

Keep current ordering guarantees by:

* Using latch or sequence tracking
* Ensuring results are emitted in input order


### 6. Buffer Size Semantics (No Breaking Change)

* Keep `bufferSize` parameter
* Update semantics:

> bufferSize controls **how many completed results can be queued**, not how many fibers can run

* Update docs to reflect this clearly

## Tests

### 1. Parallelism Not Capped

```scala
mapZIOPar(32, bufferSize = 8)
```

→ Should achieve ~32 concurrent executions

### 2. Ordering Preserved

* Output order must match input


### 3. Backpressure Safety

* Slow consumer should:

  * Not cause unbounded memory growth
  * Still limit upstream pulling


### 4. Large Parallelism Safety

* `parallelism = 10000`
* No OOM
* No deadlock


### 5. Partial Consumption

```scala
.take(1)
```

* Should not leak fibers or hang


## Non-Goals (for this PR)

* No API changes
* No removal of `bufferSize`
* No introduction of unordered variant


## Expected Outcome

* `parallelism` is fully honored
* `bufferSize` no longer caps concurrency
* Backpressure and ordering remain intact

**Comment by orbitwebsites-cloud** [comment]
/attempt #9339

**Implementation plan:**
Reimplement `ZStream#mapZIOPar` to decouple parallelism level from `bufferSize` or provide a more sensible default when `bufferSize` is not provided, ensuring the provided parallelism level is honoured.

Will submit a PR shortly.

**Comment by grp06** [comment]
/attempt #9339

Plan: fix ordered mapOutZIOPar so the ordered fiber handoff is lazily bounded by max(n, bufferSize), instead of using a preallocated bounded queue of bufferSize. This keeps n as the concurrency limit, avoids allocating n slots for Int.MaxValue, preserves ordering, and keeps slow downstream consumers from allowing unbounded upstream pull. I will add regression coverage for n > bufferSize and for fast effects under slow downstream backpressure.
